third_party/nvfuser/csrc/executor.cpp

```diff
@@ -1111,11 +1109,67 @@ std::vector<at::Tensor> FusionExecutor::runFusion(
     launch_params_ =
         computeLaunchParams(launch_constraints, expr_eval, warp_size_);
 
+    auto alias_indices_entry =
+        executor_utils::caching::ExecutorCompileTimeEntry<
+            executor_utils::caching::InputAliasIndices>(
+            compileTimeDataCache(), [&]() {
+              return std::make_unique<std::vector<std::pair<int, int>>>(
+                  fusion_->getInputAliasIndices());
+            });
+
+    auto& alias_indices = alias_indices_entry.get();
+
+    // NOLINTNEXTLINE(bugprone-branch-clone)
+    if (outputs.empty()) {
+      auto output_alias_indices_entry =
+          executor_utils::caching::ExecutorCompileTimeEntry<
+              executor_utils::caching::OutputAliasIndices>(
+              compileTimeDataCache(), [&]() {
+                return std::make_unique<std::unordered_set<int>>(
+                    fusion_->getOutputAliasIndices());
+              });
+
+      auto& output_alias_indices = output_alias_indices_entry.get();
+
+      allocated_outputs = allocOutputs(args, expr_eval, output_alias_indices);
+
+      for (const auto& entry : alias_indices) {
+        auto aliased_output_index = entry.first;
+        auto aliased_input_index = entry.second;
+        auto tensor_arg_abstract =
+            dynamic_cast<const TensorArgAbstract*>(args[aliased_input_index]);
+        TORCH_INTERNAL_ASSERT(
+            tensor_arg_abstract, "alias io only supports tensor");
+        allocated_outputs[aliased_output_index] =
+            tensor_arg_abstract->getTensor();
+      }
+      args.push(allocated_outputs);
+    } else {
+      allocated_outputs = outputs;
+      args.push(outputs);
+      executor_utils::validateKernelOutputs(
+          fusion_, allocated_outputs, options_.device);
+    }
+
     // Recompile the kernel if the number of threads in the block has increased
-    // or maxrregcount has changed
-    if (launch_params_.nThreads() > block_size_high_water_mark ||
-        compile_params.maxrregcount != maxrregcount_high_water_mark) {
+    // or maxrregcount has changed or nvfuser_index_t size changed
+    bool need_to_recompile =
+        launch_params_.nThreads() > block_size_high_water_mark ||
+        compile_params.maxrregcount != maxrregcount_high_water_mark;
+
+    // we recompile if the index type shrinks too, as it may lead to faster
+    // code.
+    if (args.getIndexMode() != options_.index_mode) {
```

Comment:
Couldn't this cause compiling back and forth if the inputs to this kernel swap back and forth from int32 representable and int64 representable?

Reply:
Yes, I'm not sure how to deal with that. Do we compile both versions and keep both?  Using KernelDb does the trick

Reply:
This is what I want to get feedback on from @naoyam and @jjsjann123. Easiest mechanism would be to actually leave the executor as is and put it as part of the heuristics choice. I'm not sure what the right call is here.

Reply:
It seems to make most sense to me if the indexing mode is by default 64 bits and can optionally be 32 bits as part of the scheduling parameters. This should allow us to cache both 32-bit and 64-bit versions.

Reply:
scheduling parameters sounds about right. IIUC, we'll still be able to reject 64-bit kernel on 32-bit representable tensor inputs/outputs.

Reply:
When you talk about scheduling parameters, do you refer to CompileParams? Should I trigger an assertion when tensors are 64b, and kernel was compiled for 32?

Reply:
I do think that's a good place. But we also need to do a little extra there. Looking at this PR we are treating index_mode similarly to how we recompile for high_water_mark register. Where we keep recompiling whenever that changes.

That's ok for high_water_mark register, since that only goes up monotonically. But in the case of indexing, we can go back and forth.

We need to cache the kernel selection here and re-use compiled kernel for alternating index mode. Which means we need to plumb it all the way to `scheduler`?!

Reply:
Right now we only compare scheduler when choosing the cached kernel: 
https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/kernel_cache.cpp#L348

https://github.com/csarofeen/pytorch/blob/4d990dc10e04213a469b1294c243439c5f54fb11/third_party/nvfuser/csrc/scheduler/registry.h#L216-L220

We need to also compare and cached that along with index_mode_.

Ah, the code in `kernel_cache.cpp` looks messy. There's some duplication in `compileKernel` and `runKernelWithInput`

Reply:
No, `CompileParams` isn't used to determine cache hits of compiled kernels (AFAIK).

It turned out that the scheduler cache actually takes care of different indexing modes, so repeated recompilation shouldn't happen as long as the fusion is compiled through the schedulers.

https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/scheduler/registry.cpp#L1037-L1039

So, I think this PR change is sufficient. We should just make sure the caching behavior is working as intended. 

@mmigdal-nv Can you add a C++ test that verifies the caching and compilation behavior? Just create a simple fusion and run it with `runFusionWithInputs` and use two sets of inputs, one that should use int32 and another that should use int64. Verify that the fusion cache creates two compiled instances and the compilation only happens two times.

Reply:
I commented above before reading @jjsjann123's comments. @jjsjann123 , am I understanding the caching scheme correctly?

Reply:
I think you are right, at least that should be the design. The sameAs function does take all those heuristics into account already. That happens at KernelRuntime.

So we should have this properly handled already.
https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/kernel_cache.cpp#L776
https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/fusion_segmenter.cpp#L2042-L2043
https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/scheduler/registry.cpp#L2139-L2140

I guess the fact that we are patching it here means something is broken in that chain?!

Reply:
I thought the fix in this PR is to make sure the executor recompiles the kernel if the index mode is changed. I believe this should never happen if compiled through the schedulers, but for matmuls we don't have a scheduler yet, so I suppose Michel is testing a manually scheduled kernel with different inputs, which could hit the error.

Reply:
@naoyam you're deduction is spot on - I create the fusion and manually schedule the matmul. Then I end-up calling `compileFusion({inputs)}` and `runFusion({inputs})`.

Reply:
So if the caching should be handled outside of fusion executor - we should never be hitting the case where runFusion switches back and forth between different types, right?

Reply:
> we should never be hitting the case where runFusion switches back and forth between different types, right?

Yes, as long as a fusion is executed through `FusionExecutorCache`.